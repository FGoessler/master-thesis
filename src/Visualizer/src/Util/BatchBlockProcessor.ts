type Placeholder = null;

// TODO: unit test this!

export default class BatchBlockProcessor<T> {

  private dataFetchingAction: (blockNumber: number) => Promise<T>;
  private dataUpdateCallback?: (datapoints: T[], labels: string[]) => void;

  constructor(
    dataFetchingAction: (blockNumber: number) => Promise<T>,
    dataUpdateCallback?: (datapoints: T[], labels: string[]) => void
  ) {
    this.dataFetchingAction = dataFetchingAction;
    this.dataUpdateCallback = dataUpdateCallback;
  }

  public async startLoading(
    fromBlockNumber: number,
    count: number,
    batchSize: number = 100
  ) {
    let datapoints = [] as T[];
    let labels = [] as string[];
    let nextBlockNumber = fromBlockNumber;

    while (nextBlockNumber > fromBlockNumber - count && nextBlockNumber >= 0) {
      let data = await this.loadBatch(nextBlockNumber, batchSize);
      datapoints = datapoints.concat(data.datapoints);
      labels = labels.concat(data.labels);
      this.sendUpdate(datapoints, labels);
      nextBlockNumber = nextBlockNumber - batchSize;
    }

    return {datapoints, labels};
  }

  private async loadBatch(fromBlockNumber: number, count: number) {
    let datapoints = [] as (T | Placeholder)[];
    let labels = [] as string[];
    let promises = [];
    let nextBlockNumber = fromBlockNumber;
    while (nextBlockNumber > fromBlockNumber - count && nextBlockNumber >= 0) {
      promises.push(this.loadDataAt(nextBlockNumber, datapoints, labels));
      nextBlockNumber = nextBlockNumber - 1;
    }
    await Promise.all(promises);
    return {datapoints: datapoints as T[], labels};
  }

  private async loadDataAt(
    blockNumber: number,
    targetDataArray: (T | Placeholder)[], targetLabelArray: string[]
  ) {
    const targetIndex = targetDataArray.length;

    targetDataArray.push(null);   // push placeholder value
    targetLabelArray.push(`${blockNumber}`);

    targetDataArray[targetIndex] = await this.dataFetchingAction(blockNumber);
  }

  private sendUpdate(datapoints: T[], labels: string[]) {
    if (this.dataUpdateCallback) {
      this.dataUpdateCallback(
        datapoints.slice(0, datapoints.length),
        labels.slice(0, datapoints.length),
      );
    }
  }
}
