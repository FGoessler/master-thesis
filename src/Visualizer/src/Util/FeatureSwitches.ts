
let featureSwitches = {
  SHOW_ADDRESS_BALANCE_CHART_DATA: false
};

export default class FeatureSwitchService {
  public static setFeatureSwitch(name: string, val: boolean) {
    featureSwitches[name] = val;
  }

  public static getFeatureSwitch(name: string): boolean {
    return !!featureSwitches[name];
  }
}
