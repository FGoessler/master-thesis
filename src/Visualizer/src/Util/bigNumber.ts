const BigNumberLib = require("bignumber.js");

const unitMap = {
  "wei":          "1",
  "kwei":         "1000",
  "ada":          "1000",
  "femtoether":   "1000",
  "mwei":         "1000000",
  "babbage":      "1000000",
  "picoether":    "1000000",
  "gwei":         "1000000000",
  "shannon":      "1000000000",
  "nanoether":    "1000000000",
  "nano":         "1000000000",
  "szabo":        "1000000000000",
  "microether":   "1000000000000",
  "micro":        "1000000000000",
  "finney":       "1000000000000000",
  "milliether":   "1000000000000000",
  "milli":        "1000000000000000",
  "ether":        "1000000000000000000",
  "kether":       "1000000000000000000000",
  "grand":        "1000000000000000000000",
  "einstein":     "1000000000000000000000",
  "mether":       "1000000000000000000000000",
  "gether":       "1000000000000000000000000000",
  "tether":       "1000000000000000000000000000000"
};

export interface BigNumber {
  toNumber(): number;
  toString(base: number): string;
  times(num: number | BigNumber): BigNumber;
  dividedBy(num: number | BigNumber): BigNumber;
}

export class BigNumberHelper {

  static toBigNumber(val: number | string | BigNumber = 0): BigNumber {
    if (val instanceof BigNumberLib) {
      return val as BigNumber;
    }

    if (typeof val === "string" && (val.indexOf("0x") === 0 || val.indexOf("-0x") === 0)) {
      return new BigNumberLib(val.replace("0x", ""), 16);
    }

    return new BigNumberLib((<number> val).toString(), 10);
  }

  static isBigNumber(object: {}) {
    return object instanceof BigNumberLib;
  }

  static toWei(num: number | string | BigNumber, unit: string) {
    return BigNumberHelper.toBigNumber(num).times(BigNumberHelper.getValueOfUnit(unit));
  }

  static fromWei(num: number | string | BigNumber, unit: string) {
    return BigNumberHelper.toBigNumber(num).dividedBy(BigNumberHelper.getValueOfUnit(unit));
  }

  private static getValueOfUnit(unit: string) {
    unit = unit ? unit.toLowerCase() : "ether";
    let unitValue = unitMap[unit];
    if (!unitValue) {
      throw new Error("Unit does not exist. Possible units: " + JSON.stringify(unitMap));
    }
    return new BigNumberLib(unitValue, 10);
  }
}
