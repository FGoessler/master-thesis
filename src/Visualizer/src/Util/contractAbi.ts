export const abi = [{
  "constant": false,
  "inputs": [],
  "name": "finalizeCloseThroughServiceConsumer",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "terminated",
  "outputs": [{"name": "", "type": "bool"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "serviceConsumer",
  "outputs": [{"name": "", "type": "address"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "tokenRedemptionTimeout",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [{"name": "numRequests", "type": "uint256"}, {"name": "hash", "type": "bytes32"}, {
    "name": "v",
    "type": "uint8"
  }, {"name": "r", "type": "bytes32"}, {"name": "s", "type": "bytes32"}],
  "name": "redeemToken",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [{"name": "", "type": "uint256"}],
  "name": "redeemedTokenIds",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [],
  "name": "payoutForServiceProvider",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "terminationTimeout",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [],
  "name": "closeThroughServiceProvider",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "soonExpiringTokenId",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "terminationStartedAt",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "serviceProvider",
  "outputs": [{"name": "", "type": "address"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "costPerRequest",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [{"name": "newTokenID", "type": "uint256"}],
  "name": "startTokenRedemption",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "serviceConsumerFunds",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "serviceProviderFunds",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "soonExpiringTokenIdTimeout",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "activeTokenId",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [],
  "name": "startCloseThroughServiceConsumer",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [],
  "name": "depositForServiceConsumer",
  "outputs": [],
  "payable": true,
  "type": "function"
}, {
  "inputs": [{"name": "_serviceProvider", "type": "address"}, {
    "name": "_initialTokenId",
    "type": "uint256"
  }, {"name": "_costPerRequest", "type": "uint256"}, {
    "name": "_terminationTimeout",
    "type": "uint256"
  }, {"name": "_tokenRedemptionTimeout", "type": "uint256"}], "payable": false, "type": "constructor"
}, {
  "anonymous": false,
  "inputs": [{"indexed": false, "name": "code", "type": "uint8"}],
  "name": "Error",
  "type": "event"
}, {
  "anonymous": false,
  "inputs": [{"indexed": false, "name": "sum", "type": "uint256"}],
  "name": "Redeemed",
  "type": "event"
}, {
  "anonymous": false,
  "inputs": [{"indexed": false, "name": "sum", "type": "uint256"}],
  "name": "Payout",
  "type": "event"
}];
