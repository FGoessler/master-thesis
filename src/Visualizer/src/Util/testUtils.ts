import { HTMLAttributes, ReactWrapper } from "enzyme";

/// This is a helper function to deal with the bit messy capabilities of enzyme to simulate form input element changes.
export function simulateInput(node: ReactWrapper<HTMLAttributes, {}>, input: string) {
  (node as any).node.value = input;
  node.first().simulate("change");
}

export function noOp() {
  return;
}
