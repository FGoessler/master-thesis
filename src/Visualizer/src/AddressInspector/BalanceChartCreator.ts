import AddressInfoService from "./AddressInfoService";
import BlockListService from "../BlockInspector/BlockListService";
import { BigNumberHelper } from "../Util/bigNumber";
import BatchBlockProcessor from "../Util/BatchBlockProcessor";

export interface ChartData {
  datapoints: number[];
  labels: string[];
  label: string;
  latestBlockNr: number;
  oldestBlockNr: number;
}

export class BalanceChartCreator {

  /// Can be called multiple times.
  public newChartDataCallback: (address: string, data: ChartData) => void;

  private addressInfoService: AddressInfoService;
  private blocklistService: BlockListService;

  constructor(addressInfoService: AddressInfoService, blocklistService: BlockListService) {
    this.addressInfoService = addressInfoService;
    this.blocklistService = blocklistService;
  }

  public async getChartDataStartingWithLatestBlock(address: string) {
    const latestBlock = await this.blocklistService.getBlock("latest");
    return this.getChartDataStartingWith(address, latestBlock.number);
  }

  public async getChartDataStartingWith(address: string, firstBlockNumber: number) {
    const dataFetchingAction = async (blockNumber: number) => {
      const balance = await this.addressInfoService.getBalanceAt(address, blockNumber);
      return BigNumberHelper.fromWei(balance, "nanoether").toNumber();
    };
    const dataUpdateCallback = (datapoints: number[], labels: string[]) => {
      this.newChartDataCallback(address, {
        datapoints: datapoints.slice(0, datapoints.length - 1),
        labels: labels.slice(0, datapoints.length - 1),
        label: address,
        latestBlockNr: parseInt(labels[0], 10),
        oldestBlockNr: parseInt(labels[labels.length - 1], 10)
      });
    };
    const batchProcessor = new BatchBlockProcessor<number>(dataFetchingAction, dataUpdateCallback);

    return batchProcessor.startLoading(firstBlockNumber, 2000);
  }

}
