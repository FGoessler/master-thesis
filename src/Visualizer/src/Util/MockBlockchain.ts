
import { Block, Eth, Web3, Transaction, TransactionReceipt } from "./web3";
import { BigNumber, BigNumberHelper } from "./bigNumber";

export class MockBlock implements Block {
  number: number = 1;
  hash: string = "0x2a";
  parentHash: string = "0x1a";
  timestamp: number = 100000;
  transactions: (string | Transaction)[] = [];

  constructor(num: number = 1) {
    this.number = num;
  }
}

export class MockTransaction implements Transaction {
  hash: string = "0xffff1a";
  from: string = "0xdd1a";
  to: string  = "0xdd2b";
  value: BigNumber = BigNumberHelper.toBigNumber(0);
  input: string = "";
  transactionReceipt?: TransactionReceipt;
  isContractDeployment: boolean = false;
}

export default class MockWeb3 implements Web3 {

  public eth: Eth;

  constructor() {
    this.eth = new MockEth();
  }
}

export class MockEth implements Eth {

  public transactionReceipts: { [key: string]: TransactionReceipt };
  public blocks: Block[];
  public balances: {};

  constructor() {
    this.transactionReceipts = {
      "0xffff1a": {
        blockHash: "0x2a",
        blockNumber: 1,
        transactionHash: "0xffff1a",
        gasUsed: 100000,
        logs: []
      }
    };

    this.blocks = [
      {
        number: 0,
        hash: "0x1a",
        parentHash: "",
        timestamp: 100000,
        transactions: []
      },
      {
        number: 1,
        hash: "0x2a",
        parentHash: "0x1a",
        timestamp: 100000,
        transactions: [
          new MockTransaction()
        ]
      },
      {
        number: 2,
        hash: "0x3a",
        parentHash: "0x2a",
        timestamp: 100000,
        transactions: []
      }
    ];

    this.balances = {
      "0x1234": BigNumberHelper.toBigNumber("9900000")
    };
  }

  contract(abi: any[]) {
    return {};
  }

  getBlock(hash: number | string, inclTransactions: boolean, cb: (err?: Error, b?: Block) => void): void {
    if (typeof hash === "number") {
      cb(undefined, this.blocks[hash]);
    } else {
      if (hash === "latest") {
        cb(undefined, this.blocks[this.blocks.length - 1]);
      } else {
        const foundBlock = this.blocks.find((b) => { return b.hash === hash; });
        if (foundBlock) {
          cb(undefined, foundBlock);
        } else {
          cb(new Error(`Block ${hash} not found!`), undefined);
        }
      }
    }
  }

  getTransactionReceipt(hash: string, cb: (err?: Error, b?: TransactionReceipt) => void): void {
    if (this.transactionReceipts[hash]) {
      cb(undefined, this.transactionReceipts[hash]);
    } else {
      cb(new Error(`Transaction ${hash} not found!`), undefined);
    }
  }

  getBalance(address: string, cb: ((err?: Error, b?: BigNumber) => void)): void;
  getBalance(address: string, block: number | string, cb: (err?: Error, b?: BigNumber) => void): void;
  getBalance(address: string,
             blockOrCb: number | string | ((err?: Error, b?: BigNumber) => void),
             cb?: ((err?: Error, b?: BigNumber) => void)): void {
    let callback: (err?: Error, b?: BigNumber) => void;
    if (typeof blockOrCb === "function") {
      callback = blockOrCb;
    } else if (cb && typeof cb === "function") {
      callback = cb;
    } else {
      throw new Error("Missing callback!");
    }

    const balance = this.balances[address];
    if (balance) {
      callback(undefined, balance);
    } else {
      callback(new Error(`Address ${address} not found!`), undefined);
    }
  }
}
