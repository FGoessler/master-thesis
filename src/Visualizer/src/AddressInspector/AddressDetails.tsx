import * as React from "react";
import StringFormatter from "../Util/stringFormatter";
import { BigNumber } from "../Util/bigNumber";
import { Line } from "react-chartjs-2";
import { ChartData } from "./BalanceChartCreator";
import { ContractValue } from "./AddressInfoService";
import ContractData from "./ContractData";
import FeatureSwitchService from "../Util/FeatureSwitches";

const chartConfig = {
  labels: [] as string[],
  datasets: [
    {
      label: "",
      fill: false,
      lineTension: 0.1,
      backgroundColor: "rgba(75,192,192,0.4)",
      borderColor: "rgba(75,192,192,1)",
      borderCapStyle: "butt",
      borderDash: [],
      borderDashOffset: 0.0,
      borderJoinStyle: "miter",
      pointBorderColor: "rgba(75,192,192,1)",
      pointBackgroundColor: "#fff",
      pointBorderWidth: 1,
      pointHoverRadius: 5,
      pointHoverBackgroundColor: "rgba(75,192,192,1)",
      pointHoverBorderColor: "rgba(220,220,220,1)",
      pointHoverBorderWidth: 2,
      pointRadius: 1,
      pointHitRadius: 10,
      data: [] as number[]
    }
  ]
};

interface AddressDetailsProps {
  address: string;
  currentBalance?: BigNumber;
  chartData: ChartData;
  contractData?: ContractValue[];
  onRequestOlderBlocks: (address: string, nextBlockNr: number) => void;
}

export default function AddressDetails(props: AddressDetailsProps) {

  let chart = undefined;
  if (FeatureSwitchService.getFeatureSwitch("SHOW_ADDRESS_BALANCE_CHART_DATA")) {
    if (props.chartData.datapoints.length > 0) {
      const config = Object.assign({}, chartConfig);
      config.labels = props.chartData.labels;
      config.datasets[0].label = props.chartData.label;
      config.datasets[0].data = props.chartData.datapoints;
      chart = (
        <div>
          <h4>Balance Chart</h4>
          <Line data={config}/>
          <button onClick={() => props.onRequestOlderBlocks(props.address, props.chartData.oldestBlockNr - 1)}>
            Load data for older blocks
          </button>
        </div>
      );
    } else {
      chart = <span>Chart loading...</span>;
    }
  }

  return (
    <div>
      <h3>
        {StringFormatter.trimHash(props.address)}
      </h3>
      <table>
        <tbody>
        <tr>
          <td>Address</td>
          <td className="hash">{StringFormatter.trimHash(props.address)}</td>
        </tr>
        <tr>
          <td>Balance</td>
          <td>{props.currentBalance ? props.currentBalance.toString(10) : "..."}</td>
        </tr>
        </tbody>
      </table>
      <ContractData contractData={props.contractData}/>
      {chart}
    </div>
  );
}
