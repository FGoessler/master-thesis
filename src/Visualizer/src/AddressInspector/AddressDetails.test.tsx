import * as React from "react";
import * as ReactDOM from "react-dom";
import AddressDetails from "./AddressDetails";
import { BigNumberHelper } from "../Util/bigNumber";
import { ChartData } from "./BalanceChartCreator";
import { noOp } from "../Util/testUtils";
import { mount } from "enzyme";
import FeatureSwitchService from "../Util/FeatureSwitches";

const dummyChartData = {datapoints: [], label: "", labels: [], oldestBlockNr: 0, latestBlockNr: 0} as ChartData;

it("shows information about an address", async () => {
  const item = document.createElement("div");
  ReactDOM.render(
    <AddressDetails
      address={"0x123456789abcde"}
      currentBalance={BigNumberHelper.toBigNumber(10000)}
      chartData={dummyChartData}
      onRequestOlderBlocks={noOp}
    />,
    item);
  expect(item).toMatchSnapshot();
});

it("can handle undefined balance", async () => {
  const item = document.createElement("div");
  ReactDOM.render(
    <AddressDetails
      address={"0x123456789abcde"}
      currentBalance={undefined}
      chartData={dummyChartData}
      onRequestOlderBlocks={noOp}
    />,
    item);
  expect(item).toMatchSnapshot();
});

describe("chart", () => {

  beforeEach(() => {
    FeatureSwitchService.setFeatureSwitch("SHOW_ADDRESS_BALANCE_CHART_DATA", true);
  });

  afterEach(() => {
    FeatureSwitchService.setFeatureSwitch("SHOW_ADDRESS_BALANCE_CHART_DATA", false);
  });

  it("shows a chart", async () => {
    const item = document.createElement("div");
    const chartData = {
      label: "the chart",
      datapoints: [1, 3, 6, 9], labels: ["1", "2", "3", "4"],
      oldestBlockNr: 2, latestBlockNr: 5
    };
    ReactDOM.render(
      <AddressDetails
        address={"0x123456789abcde"}
        currentBalance={undefined}
        chartData={chartData}
        onRequestOlderBlocks={noOp}
      />,
      item);
    expect(item).toMatchSnapshot();
  });

  it("propagates the click on the request older blocks button", async () => {
    const requestOlderBlocks = jest.fn();
    const chartData = {
      label: "the chart",
      datapoints: [1, 3, 6, 9], labels: ["1", "2", "3", "4"],
      oldestBlockNr: 2, latestBlockNr: 5
    };
    const list = mount(
      <AddressDetails
        address={"0x123456789abcde"}
        currentBalance={undefined}
        chartData={chartData}
        onRequestOlderBlocks={requestOlderBlocks}
      />);
    list.find("button").first().simulate("click");

    expect(requestOlderBlocks).toHaveBeenCalledWith("0x123456789abcde", chartData.oldestBlockNr - 1);
  });

});
