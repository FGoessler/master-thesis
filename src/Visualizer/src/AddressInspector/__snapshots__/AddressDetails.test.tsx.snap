// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can handle undefined balance 1`] = `
<div>
  <div
    data-reactroot=""
  >
    <h3>
      0x123456789abcde
    </h3>
    <table>
      <tbody>
        <tr>
          <td>
            Address
          </td>
          <td
            class="hash"
          >
            0x123456789abcde
          </td>
        </tr>
        <tr>
          <td>
            Balance
          </td>
          <td>
            ...
          </td>
        </tr>
      </tbody>
    </table>
    <div>
      No Contract Data
    </div>
  </div>
</div>
`;

exports[`chart shows a chart 1`] = `
<div>
  <div
    data-reactroot=""
  >
    <h3>
      0x123456789abcde
    </h3>
    <table>
      <tbody>
        <tr>
          <td>
            Address
          </td>
          <td
            class="hash"
          >
            0x123456789abcde
          </td>
        </tr>
        <tr>
          <td>
            Balance
          </td>
          <td>
            ...
          </td>
        </tr>
      </tbody>
    </table>
    <div>
      No Contract Data
    </div>
    <div>
      <h4>
        Balance Chart
      </h4>
      <canvas
        height="150"
        width="300"
      />
      <button>
        Load data for older blocks
      </button>
    </div>
  </div>
</div>
`;

exports[`shows information about an address 1`] = `
<div>
  <div
    data-reactroot=""
  >
    <h3>
      0x123456789abcde
    </h3>
    <table>
      <tbody>
        <tr>
          <td>
            Address
          </td>
          <td
            class="hash"
          >
            0x123456789abcde
          </td>
        </tr>
        <tr>
          <td>
            Balance
          </td>
          <td>
            10000
          </td>
        </tr>
      </tbody>
    </table>
    <div>
      No Contract Data
    </div>
  </div>
</div>
`;
